source/common/common/compiled_string_map.h

```diff
@@ -0,0 +1,157 @@
+#pragma once
+
+#include <algorithm>
+#include <string>
+#include <vector>
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+
+/**
+ * This is a specialized structure intended for static header maps, but
+ * there may be other use cases.
+ * The structure is:
+ * 1. a length-based lookup table so only keys the same length as the
+ * target key are considered.
+ * 2. a trie that branches on the "most divisions" position of the key.
+ *
+ * For example, if we consider the case where the set of headers is
+ * `x-prefix-banana`
+ * `x-prefix-babana`
+ * `x-prefix-apple`
+ * `x-prefix-pineapple`
+ * `x-prefix-barana`
+ * `x-prefix-banaka`
+ *
+ * A standard front-first trie looking for `x-prefix-banana` would walk
+ * 7 nodes through the tree, first for `x`, then for `-`, etc.
+ *
+ * This structure first jumps to matching length, eliminating in this
+ * example case apple and pineapple.
+ * Then the "best split" node is on
+ *   `x-prefix-banana`
+ *               ^
+ * so the first node has 3 non-miss branches, n, b and r for that position.
+ * Down that n branch, the "best split" is on
+ *   `x-prefix-banana`
+ *                 ^
+ * which has two branches, n or k.
+ * Down the n branch is the leaf node (only `x-prefix-banana` remains) - at
+ * this point a regular string-compare checks if the key is an exact match
+ * for the string node.
+ */
+template <class Value> class CompiledStringMap {
+  using FindFn = std::function<Value(const absl::string_view&)>;
+
+public:
+  using KV = std::pair<std::string, Value>;
```

Comment:
I made the input a vector of string_views which eliminates the string copying during compile() except the copy taken for a leaf node (which is kept in case at some point the table wants to be initialized from a transient source rather than global singleton-constant-likes).